pkg/apis/experimental/types.go

```diff
@@ -461,3 +461,75 @@ type JobCondition struct {
 	// Human readable message indicating details about last transition.
 	Message string `json:"message,omitempty"`
 }
+
+// IngressPoint encapsulates the inputs needed to connect an alias to endpoints.
+// It represents a rule collection of inbound connections from the external network
+// that would be satisfied by a load balancer.
+type IngressPoint struct {
+	api.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the IngressPoint.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Spec IngressPointSpec `json:"spec,omitempty"`
+
+	// Status is the current state of the IngressPoint.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Status IngressPointStatus `json:"status,omitempty"`
+}
+
+// IngressPointList is a collection of IngressPoints,
+type IngressPointList struct {
+	api.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	api.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of ingressPoints.
+	Items []IngressPoint `json:"items"`
+}
+
+// ServiceRef represents a reference to a single service:port.
+type ServiceRef struct {
+	// Define the name of the referenced service.
+	Name string `json:"name,omitempty"`
+
+	// Define the namespace of the referenced service.
+	Namespace string `json:"namespace,omitempty"`
+
+	// Define the service's port which the http traffic will be forwarded to.
+	Port int `json:"port,omitempty"`
+}
+
+// PathRef represents a single load balancer rule of forwarding the specified url to the related backend service.
+type PathRef struct {
+	// Domain describe http url's domain.
+	// Examples: "foo.example.com", "bar.example.com".
+	Domain string `json:"domain,omitempty"`
```

Comment:
Some changes:
```
// This is just like the LocalObjectReference but includes namespace
NonLocalObjectReference {
  Name string
  Namespace string
}

Ingress {
  Spec IngressSpec
  Status IngressStatus
}

IngressStatus {
  // This is an existing struct
  LoadBalancers LoadBalancerStatus
}

IngressSpec {
    MatchingDomain string,
    MatchingSubDomains []IngressMatchingSubDomain
}

IngressMatchingSubDomain {
    SubDomain string
    MatchingPathExpr []IngressMatchingPathExpr
}

IngressMatchingPathExpr {
    PathExpr string
    TargetService IngressTargetService
}

IngressTargetService struct {
    ServiceRef NonLocalObjectReference
    ServicePort int
}

```
Essentially:
* The "Ingress" prefix denotes that the struct is nested in the Ingress top level resource (examples: ServicePort, PodStatus etc). 
* The "Matching" prefix denotes what parts of the resource are used to perform the match. 
* The "Target" prefix denotes what parts of the resource are the target of the match. 
* "Expr" for expression, denotes a non-exact match. 
* Lists have an "s" suffix.